txpipe/noise_maps.py

```diff
@@ -541,3 +541,216 @@ def finalize_mappers(self, pixel_scheme, mappers):
+        counter = 0
+        for i in range(pixel_scheme.npix):
+            if mask[i] > 0:
+                index_map[i] = counter
+                counter += 1
+        # Number of unmasked pixels
+        npix = counter
+
+        if self.rank == 0:
+            nmaps = nbin_source * (2 * lensing_realizations + 1) + nbin_lens * clustering_realizations * 2
+            nGB = (npix * nmaps * 8) / 1000.**3
+            print(f"Allocating maps of size {nGB:.2f} GB")
+        # lensing g1, g2
+        G1 = jnp.zeros((npix, nbin_source, lensing_realizations))
+        G2 = jnp.zeros((npix, nbin_source, lensing_realizations))
+        # lensing weight
+        GW = jnp.zeros((npix, nbin_source))
+        # clustering map - n_gal to start with
+        ngal_split = jnp.zeros((npix, nbin_lens, clustering_realizations, 2), dtype=np.int32)
+        # TODO: Clustering weights go here
+
+        # Initialize PRNG key for Jax
+        key = random.PRNGKey(np.random.randint(2**32))
+
+        # Loop through the data
+        for (s, e, data) in it:
+            print(f"Rank {self.rank} processing rows {s} - {e}")
+            # Send data to GPU
+            data = device_put(data)
+            source_bin = data['source_bin']
+            lens_bin = data['lens_bin']
+            ra = data['ra']
+            dec = data['dec']
+            orig_pixels = device_put(pixel_scheme.ang2pix(ra, dec))
+            pixels = device_put(index_map[orig_pixels])
+            n = e - s
+
+            weights = data['weight']
+            g1 = data['mcal_g1'] * weights
+            g2 = data['mcal_g2'] * weights
+            # randomly select a half for each object
+            key, subkey = random.split(key)
+            split = 1*random.bernoulli(subkey, 0.5, (n, clustering_realizations))
+
+            # random rotations of the g1, g2 values
+            key, subkey = random.split(key)
+            phi = random.uniform(subkey, shape=(lensing_realizations, n), minval=0, maxval=2*jnp.pi)
+            cos = jnp.cos(phi)
+            sin = jnp.sin(phi)
+            g1r = jnp.transpose(cos * g1 + sin * g2)
+            g2r = jnp.transpose(-sin * g1 + cos * g2)
+
+            pix_mask = pixels >= 0
+            sb_mask = (source_bin >= 0) & pix_mask
+            G1 = G1.at[index[pixels[sb_mask], source_bin[sb_mask], :]].add(g1r[sb_mask])
+            G2 = G2.at[index[pixels[sb_mask], source_bin[sb_mask], :]].add(g2r[sb_mask])
+            GW = GW.at[index[pixels[sb_mask], source_bin[sb_mask]]].add(weights[sb_mask])
+            lb_mask = sb_mask & (lens_bin >= 0)
+            ngal_split = ngal_split.at[index[pixels[lb_mask], lens_bin[lb_mask], jnp.arange(clustering_realizations), split[lb_mask]]].add(1)
+            #!!!!! Currently breaks with clustering_realizations > 1
+        # Sum everything at root
+        if self.comm is not None:
+            import mpi4jax
+            from mpi4py import MPI
+            mpi4jax.reduce(G1, MPI.SUM, root=0)
```

Comment:
I think root=0 is the default? Fine to leave it if it's for clarity though.

Reply:
I think that's true in mpi4py but not sure in mpi4jax [https://mpi4jax.readthedocs.io/en/latest/api.html?highlight=reduce#mpi4jax.reduce](url), at least they show default values for `comm` and `token` but not `root`